\section{Introduction}

Sequence data can be stored as text with each letter representing a nucleic acid (DNA and RNA) or amino acid (protein). The linear nature of these molecules makes it natural to represent them as strings, finite sequence of characters. Although it has been argued that a graph, a network of edges connected by vertices, is actually a more accurate way to store genomic sequences because graphs allow for the inclusion of alternate alleles and alternate possible assemblies \cite{jaffe2012fastg} all of the most common methods for storing sequences (FASTA, FASTQ, SAM/BAM) use a linear strings.

Other decisions about how to represent sequence data can be more arbitrary. For example, any character that is not used as base or an amino acid can be used to indicate the beginning of a new sequence. Additionally text can be wrapped to limit the information content in any one line of a file. The advantage of wrapping text is that some programs can then be designed to work one line at time limiting the burden of each step (e.g. the program would never have to process an entire chromosome of sequence data in a single step). The disadvantage is that code must be slightly more complex to load an entire sequence record into the working memory.

\subsection{FASTA file format specifications versus recommendations}

FASTA file format requirements are very minimal \cite{FASTAformat}. Each sequence is preceded by a header/description line that begins with a \verb|>|. Sequence lines can include any standard IUB/IUPAC single character symbols for nucleic acids or amino acids or the ambiguous codes that indicate possible residues or bases \cite{comm1970abbreviations}. They can also include \verb|-| to indicate alignment gaps and \verb|*| to indicate stop codons. 

NCBI recommends wrapping FASTA file sequences lines \cite{FASTAformat}. It is also common practice to use the first `word' in a header (i.e. any character string to the left of the first space in the header) as the unique sequence id. Although these features are common they are not required leading to format compatibility issues with tools that treat these conventions as required features.

\subsection{Customizing FASTA files to ensure that information is properly interpreted by downstream tools}

Regardless of whether a FASTA file is technically improperly formatted or it's format merely violates a popular convention it is critical to quality analysis workflows that data is converted into a format that will be correctly interpreted by downstream tools. Formatting issues can fall into multiple categories including actual format errors, formats that are not technically wrong but are non-standard and formats that throw errors because an existing tool has a bug (in which case we should modify the FASTA and proceed only if the tool will then correctly import the data and export the desired output)). 

Some format errors are more often indicative of a major problem like an attempt to use the wrong data format (e.g. the first line is not a FASTA header because it does not begin with a \verb|>| character). These types of errors will be subsequently referred to as fatal. However some formatting issues are common and typically do not indicate the input file is corrupt (e.g. improperly wrapped/ unwrapped sequence lines, missing final new line characters, unusual new line characters like \verb|\r|). These issues will be referred to as non-fatal. Fatal formatting issues should cause processing to stop. Non-fatal formatting issues should be automatically corrected according to the most common resolution for this type of error. While downstream processing continues the analyst can double check the automated decision to reformat non-fatal issues. In this manner the workflow would not need to be slowed for trivial reformatting steps and the more rare problems (e.g. when a missing last new line was caused by incomplete file transfer) can still be caught.
  
\subsection{Existing tools}

Many bioinformatics tools already address FASTA format inconsistencies. However all existing tools identified by the authors halt and exit with and error (e.g. BioPerl or Sequel) and/or produce output FASTA regardless of whether a formatting issue was detected (e.g. EMBOSS Secret). For example, EMBOSS Secret was designed as a very flexible tool to convert from one properly formatted file to another properly but distinctly formatted file. It also was designed to accept poorly formatted data (e.g. a FASTA missing the final new line that is improperly wrapped) and export a reformatted file (e.g. wrapped after 60 bases with a final newline). After submitting this inconsistently wrapped FASTA record that is missing a final new line character, seqret (http://www.ebi.ac.uk/Tools/sfc/emboss_seqret/) produced a properly formatted FASTA record.

Code:

\verb|seqret -auto -stdout -sequence emboss_seqret-I20150716-200022-0179-11804058-oy.sequence -snucleotide1 -sformat1 pearson -osformat2 fasta -feature -ofname2 emboss_seqret-I20150716-200022-0179-11804058-oy.gff|

Input:

\verb|>my header|

\verb|AAAAAAAAAAAATTTTTTCCCCGGCGCGCGCGCTATAGCGCTATANNNNNNNNNNNNNNN|

\verb|ATATATATATAT|

\verb|ATTATTATATATATATTCTCTCTGGGCTCGCGTCTCGCTATTTATATATATATATATATTGCGCTCTCGTCTCCT|

Output:

\verb|>my header|

\verb|AAAAAAAAAAAATTTTTTCCCCGGCGCGCGCGCTATAGCGCTATANNNNNNNNNNNNNNN|

\verb|ATATATATATATATTATTATATATATATTCTCTCTGGGCTCGCGTCTCGCTATTTATATA|

\verb|TATATATATATTGCGCTCTCGTCTCCT|

However, Seqret did not log the detected errors in the format. In many cases missing new line characters at the end of a file or variable line wrapping do not indicate corrupted data. However they could be the result of a corrupted file and the analyst should be made aware of such errors to briefly investigate the issue. Another feature of Seqret is that an output file is created even if the output is identical to the input. Storing two identical files is an inefficient use of disk space. Rather a tool should test for proper file format and export a reformatted file only if the input is found to have a formatting issue. Another example of a tool that can automate FASTA reformatting is Seqtk. The free tool Seqtk can wrap an improperly wrapped FASTA file to a user specified length but this tool was not developed to first test if wrapping was needed or already a feature of the input format. This tool has both limited functionality and would double the disk space required to store FASTA files in a workflow. Storage is trivial unless the FASTA files in question store whole genomes in which case the burden can add up for a bioinformatics core. Overall while many tools can either detect a format issue or repair a format issue no existing tool was found that both validates FASTA format and reformats automatically only where required for a user defined list of FASTA format issues.

  
  
  
  
  
  
  
  
  
  
  
  
  